pinot-transport/src/main/java/com/linkedin/pinot/routing/HelixExternalViewBasedRouting.java

```diff
@@ -176,76 +196,406 @@ public void shutdown() {
+      List<InstanceConfig> instanceConfigs, List<ServerToSegmentSetMap> serverToSegmentSetMaps) {
+    Set<String> relevantInstanceNames = new HashSet<>();
+
+    // Gather all the instance names contained in the routing table
+    for (ServerToSegmentSetMap serverToSegmentSetMap : serverToSegmentSetMaps) {
+      relevantInstanceNames.addAll(serverToSegmentSetMap.getServerSet());
+    }
+
+    // Update the relevant instance config map with the instance configs given
+    for (InstanceConfig instanceConfig : instanceConfigs) {
+      if (relevantInstanceNames.contains(instanceConfig.getInstanceName())) {
+        relevantInstanceConfigs.put(instanceConfig.getInstanceName(), instanceConfig);
+      }
+    }
   }
 
   public void markDataResourceOffline(String tableName) {
     LOGGER.info("Trying to remove data table from broker for {}", tableName);
     _brokerRoutingTable.remove(tableName);
-    _routingTableLastKnownZkVersionMap.remove(tableName);
+    _lastKnownExternalViewVersionMap.remove(tableName);
+    _lastKnownInstanceConfigsForTable.remove(tableName);
     _timeBoundaryService.remove(tableName);
+
+    // Remove table from all instances
+    synchronized (_tablesForInstance) {
+      for (String instanceName : _brokerRoutingTable.keySet()) {
+        Set<String> tablesForCurrentInstance = _tablesForInstance.get(instanceName);
+        if (tablesForCurrentInstance.contains(tableName)) {
+          tablesForCurrentInstance.remove(tableName);
+        }
+
+        if (tablesForCurrentInstance.isEmpty()) {
+          _lastKnownInstanceConfigs.remove(instanceName);
+        }
+      }
+    }
+  }
+
+  public void processExternalViewChange() {
+    System.out.println("--------------------------------");
+    System.out.println("HelixExternalViewBasedRouting.processExternalViewChange");
+    System.out.println("--------------------------------");
+
+    long startTime = System.currentTimeMillis();
+    // Get stats for all tables that we're serving
+    HelixDataAccessor helixDataAccessor = _helixManager.getHelixDataAccessor();
+    PropertyKey.Builder propertyKeyBuilder = helixDataAccessor.keyBuilder();
+    List<String> tablesServed = new ArrayList<>(_lastKnownExternalViewVersionMap.keySet());
+    List<String> externalViewPaths = new ArrayList<>(tablesServed.size());
+
+    for (String tableName : tablesServed) {
+      PropertyKey propertyKey = propertyKeyBuilder.externalView(tableName);
+      externalViewPaths.add(propertyKey.getPath());
+    }
+
+    long initTime = System.currentTimeMillis();
+
+    Stat[] externalViewStats =
+        helixDataAccessor.getBaseDataAccessor().getStats(externalViewPaths, AccessOption.PERSISTENT);
+
+    long statsTime = System.currentTimeMillis();
+
+    // Make a list of external views that changed
+    List<String> tablesThatChanged = new ArrayList<>();
+
+    for (int i = 0; i < externalViewStats.length; i++) {
+      Stat externalViewStat = externalViewStats[i];
```

Comment:
It would indicate in this case that the external view got deleted for some reason. `ZooKeeper#exists` can return null, see http://zookeeper.apache.org/doc/r3.4.6/api/org/apache/zookeeper/ZooKeeper.html#exists(java.lang.String, org.apache.zookeeper.Watcher)